### ISSUE TITLE
Calculation of split gains

### ISSUE THREAD

**Issue by mavillan** [microsoft/LightGBM#5986]
Hi all, 

The split gain as defined here: 
https://github.com/microsoft/LightGBM/blob/c870c57a14b61329bf493b3b1a9a2893a62405cf/src/treelearner/feature_histogram.hpp#L714-L746

and here:
https://github.com/microsoft/LightGBM/blob/c870c57a14b61329bf493b3b1a9a2893a62405cf/src/treelearner/feature_histogram.hpp#L749-L764

is then calculated like this:
```math
\text{split\_gain} = \frac{\max(0, G_{\text{left}}-\alpha)^2}{H_{\text{left}}+\lambda} + \frac{\max(0, G_{\text{right}}-\alpha)^2}{H_{\text{right}}+\lambda}
```

where $H$ represents sum of hessians and $G$ sum of gradients. However, shouldn't be the gain the difference between loss in the parent node, less the sum of the losses of both child nodes (left+right)? Something like this:

```math
\begin{align}

\text{gain} &= \text{loss}_{\text{parent}} - (\text{loss}_{\text{left}} + \text{loss}_{\text{right}}) \\
&= -\frac{1}{2} 
\frac{G_{\text{parent}}^2}{H_{\text{parent}}+\lambda} - \left( -\frac{1}{2} \frac{G_{\text{left}}^2}{H_{\text{left}}+\lambda} - \frac{1}{2} \frac{G_{\text{right}}^2}{H_{\text{right}}+\lambda} \right) \\
&= \frac{1}{2} \left[ \frac{G_{\text{left}}^2}{H_{\text{left}}+\lambda} + \frac{G_{\text{right}}^2}{H_{\text{right}}+\lambda} - \frac{G_{\text{parent}}^2}{H_{\text{parent}}+\lambda} \right]  \\
&= \frac{1}{2} \left[ \underbrace{\frac{G_{\text{left}}^2}{H_{\text{left}}+\lambda}}_{\text{left gain}} + \underbrace{\frac{G_{\text{right}}^2}{H_{\text{right}}+\lambda}}_{\text{right gain}} - \frac{(G_{\text{left}}+G_{\text{right}})^2}{H_{\text{left}}+H_{\text{right}}+\lambda} \right]

\end{align}
```
(also pointed out here: https://xgboost.readthedocs.io/en/latest/tutorials/model.html#learn-the-tree-structure)

My questions are:

- Is my understanding correct?
- If so, why you decided to not include the parent term into the split gain calculation?
- Couldn't that lead to a situation were you split a leaf without generating real gains? e.g: $`\text{loss}_{\text{left}} + \text{loss}_{\text{right}} > 0`$ but. $` \text{loss}_{\text{parent}} <  \text{loss}_{\text{left}} + \text{loss}_{\text{right}} `$

**Comment by jmoralez** [comment]
Hey @mavillan, thanks for using LightGBM. The parent node's gain is included (it's computed as gain left + gain right - gain parent). You can find a verification using Python [here](https://github.com/microsoft/LightGBM/issues/4500#issuecomment-897324836).

Please let us know if you have further doubts.

**Comment by mavillan** [comment]
Hi @jmoralez thanks for your quick reply.  I played around with your verification code and it make sense now! However, I would like to know where in the C++ modules I can find that calculation. The parts of the code the I referenced doesn't seem to be considering the parent gain.

**Comment by jmoralez** [comment]
I think it's because when trying to assess if there's an improvement the comparison is gain left + gain right > gain parent, for example here: https://github.com/microsoft/LightGBM/blob/c870c57a14b61329bf493b3b1a9a2893a62405cf/src/treelearner/feature_histogram.hpp#L846-L848
where the `min_gain_shift` is computed here
https://github.com/microsoft/LightGBM/blob/c870c57a14b61329bf493b3b1a9a2893a62405cf/src/treelearner/feature_histogram.hpp#L160-L162
as the parent's gain plus the min gain to split:
https://github.com/microsoft/LightGBM/blob/c870c57a14b61329bf493b3b1a9a2893a62405cf/src/treelearner/feature_histogram.hpp#L96-L111

Hopefully @shiyu1994 can correct me if I'm wrong.

**Comment by mavillan** [comment]
Make sense to now. Thanks a lot!
